### ISSUE TITLE
Saving a composite model that includes a custom layer results in error - None has NoneType, but expected one of: bytes, unicode

### ISSUE THREAD

**Issue by EnderWiggin14** [tensorflow/tensorflow#51840]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): A custom variant of RedHat, outside of my control
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.5.0
- Python version: 3.7.2
- CUDA/cuDNN version: 11.1.0/8.1.1
- GPU model and memory: Nvidia K40 12GB

**Describe the current behavior**
I'm trying to save a model which is a composite model of composite models.

The first model is a sequential model of two sequential models. Both of the two sub-models have custom layers that perform scaling operations like MinMax and cube root. This model saves and loads without any issues.

The first model is then loaded in a different script without compiling. There is not issue with this. Let's call this model MODEL_1.

The next step may be a bit confusing. There are then two more models added in parallel to each other but sequentially with MODEL_1. Let's call these models MODEL_2a and MODEL_2b. The output of MODEL_1 has the input of MODEL_1 concatenated to it and this serves as the input to MODEL_2a and MODEL_2b. It should be noted that that the "MODEL_1 input" goes through a custom scaling layer before being concatenated to the output of MODEL_1. The is scaling layer that was also implemented in MODEL_1 without any issues.

Finally, there is a single, custom layer that performs a 'simple' weighted sum of the outputs of MODEL_2a and MODEL_2b to produce the model output. This weighting happens via `alpha*OUTPUT_2a + [1-alpha]*OUTPUT_2b`. 'alpha' is a trainable, scalar parameter. I haven't used this before in a model that I have saved, so I'm guessing this is the cause.

The model compiles and trains, but fails to save.

The custom weighted sum layer is this,

```
class WeightedSum(krs.layers.Layer):
    def __init__( self, n_models = 2, name = 'weighted_sum_0' ):
        super( WeightedSum, self ).__init__( name = name)
        self.n_models = n_models
        self.ensemble_weights = []
        self.output_init = tf.Variable(0.,validate_shape=False,trainable=False)

    def build(self,input_shape):
        for i in range(self.n_models):
            self.ensemble_weights.append( self.add_weight(shape=(1,),
                                    initializer = 'ones',
                                    trainable = True) )

    def call(self,inputs):
        new_normalizer = tf.convert_to_tensor(0.,dtype = inputs[0].dtype)
        for i in range(self.n_models):
            new_normalizer = new_normalizer + self.ensemble_weights[i]
        new_normalizer = tf.constant(1.,dtype=new_normalizer.dtype)/new_normalizer
        output = self.output_init

        for i in range(self.n_models):
            output = tf.add(output,tf.multiply(self.ensemble_weights[i],inputs[i]))
        output = tf.multiply( output, new_normalizer )
        return output
```

The save command is this, (NOTE: I use `import tensorflow.keras as krs`)
```
krs.models.save_model(linked_model,"test_failed_save.mdl")
```

The error that is produced is this.

```
Traceback (most recent call last):
  File "multi_fidelity_training_full_link.py", line 304, in <module>
    main()
  File "multi_fidelity_training_full_link.py", line 265, in main
    krs.models.save_model(linked_model,"test_failed_save.mdl")
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/keras/saving/save.py", line 151, in save_model
    signatures, options, save_traces)
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/save.py", line 90, in save
    model, filepath, signatures, options)
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py", line 1104, in save_and_return_nodes
    raise_metadata_warning))
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py", line 1291, in _build_meta_graph
    raise_metadata_warning)
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py", line 1225, in _build_meta_graph_impl
    options.namespace_whitelist)
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py", line 713, in _fill_meta_graph_def
    _call_function_with_mapped_captures, resource_map=resource_map)))
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/graph_view.py", line 424, in frozen_saveable_objects
    call_with_mapped_captures)
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/graph_view.py", line 375, in _serialize_gathered_objects
    slot_variables=slot_variables)
  File "/usr/WS2/mvander/py3venv/py3venv/lib/python3.7/site-packages/tensorflow/python/training/tracking/graph_view.py", line 355, in _fill_object_graph_proto
    child_proto.local_name = child.name
TypeError: None has type NoneType, but expected one of: bytes, unicode
```

**Describe the expected behavior**
It should save the model.

**Standalone code to reproduce the issue**
```
import numpy as np
import tensorflow.keras as krs
import tensorflow as tf




class WeightedSum(krs.layers.Layer):
    def __init__( self, n_models = 2, **kwargs):
        super( WeightedSum, self ).__init__( **kwargs)
        self.n_models = n_models
        self.ensemble_weights = []
        self.output_init = tf.Variable(0.,validate_shape=False,trainable=False)

    def build(self,input_shape):
        for i in range(self.n_models):
            self.ensemble_weights.append( self.add_weight(shape=(1,),
                                    initializer = 'ones',
                                    trainable = True) )

    def call(self,inputs):
        new_normalizer = tf.convert_to_tensor(0.,dtype = inputs[0].dtype)
        for i in range(self.n_models):
            new_normalizer = new_normalizer + self.ensemble_weights[i]
        new_normalizer = tf.constant(1.,dtype=new_normalizer.dtype)/new_normalizer
        output = tf.cast(self.output_init,dtype=inputs[0].dtype)

        for i in range(self.n_models):
            output = tf.add(output,tf.multiply(tf.cast(self.ensemble_weights[i],dtype=inputs[i].dtype),inputs[i]))
        output = tf.multiply( output, new_normalizer )
        return output


input_lf = krs.Input((4,))

x = input_lf
x = krs.layers.Dense(10,activation = 'relu')(x)
x = krs.layers.Dense(10,activation = 'relu')(x)
lf_out = krs.layers.Dense(10,activation = 'relu')(x)

lf_mod = krs.Model(input_lf,lf_out,name='lf')

input_hf_lin = krs.Input((14,))
x = input_hf_lin
x = krs.layers.Dense(10)(x)
x = krs.layers.Dense(10)(x)
hf_lin_out = krs.layers.Dense(10,activation = 'relu')(x)

hf_lin_mod = krs.Model(input_hf_lin,hf_lin_out,name='hf_linear')

input_hf_nonlin = krs.Input((14,))

x = input_hf_nonlin
x = krs.layers.Dense(10,activation = 'relu')(x)
x = krs.layers.Dense(10,activation = 'relu')(x)
hf_nonlin_out = krs.layers.Dense(10,activation = 'relu')(x)

hf_nonlin_mod = krs.Model(input_hf_lin,hf_lin_out,name='hf_nonlinear')

input_hf = krs.Input((14,))

x = input_hf
lin = hf_lin_mod(x)
nonlin = hf_nonlin_mod(x)
summed_out = WeightedSum(n_models=2)([lin,nonlin])

hf_mod = krs.Model(input_hf,summed_out,name='hf')

input_full_mod = krs.Input((4,))
x = input_full_mod

low = lf_mod(x)
x = krs.layers.Concatenate()([low,x])
full_out = hf_mod(x)

full_mod = krs.Model(input_full_mod,outputs = {'low_fidelity':low,'high_fidelity':full_out},name='full_model')

opt = krs.optimizers.Adam()
loss = krs.losses.MSE
full_mod.compile(optimizer = opt,loss = loss)

x_train = np.random.uniform(0,10,(20,4))
y_train_low = np.random.uniform(0,10,(20,10))
y_train_high = np.random.uniform(0,10,(20,10))
y = {"low_fidelity": y_train_low,
     "high_fidelity": y_train_high}



full_mod.fit(x_train,y,epochs=5)

krs.models.save_model(full_mod,"test_model.mdl")
```

EDIT: I goofed during my copy and paste from VIM. I didn't get all of the code orginally.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by EnderWiggin14** [comment]
Before someone says that I should post this to StackOverflow, I have already done this and have received no response. Also, similar issues appear to generally be bugs.

**Comment by tilakrayal** [comment]
@EnderWiggin14 ,
I was able to execute the given code without facing any errror.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/d9c1dcadfd3315cc4c7948c54635b5f7/un51840.ipynb).Can you please try to execute the code in new virtual environment and let us know if you are facing same issue.Thanks!

**Comment by EnderWiggin14** [comment]
@tilakrayal I goofed during copy-paste and missed the last bit of code. I have now edited the post which includes all of the necessary code. FYI I'm running in 2.5.0 not 2.6.0. I did install 2.6.0 to test like you suggested but it failed which is how I realized i failed to do a proper copy-paste. I will be reverting to 2.5.0 because I can't justify moving to 2.6.0 as I don't have control over what CUDA/cuDNN versions are available to me on servers.

**Comment by EnderWiggin14** [comment]
@tilakrayal I believe I have found the issue. The issue is that I do not have the "name" parameter set in `self.add_weight`. 

I have no idea if it was intended by the developers to absolutely need a name provided, but the [documentation](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#add_weight) doesn't make that clear at all. It also doesn't follow the general convention within the Tensorflow API where names are optional (or at least appear to be) and default to `None`.

You'll also notice that I don't have a `get_config` implementation in the post, but that is a minor detail and is fixed in my actual implementation.

**Comment by tilakrayal** [comment]
@EnderWiggin14 ,
 
Please post this issue on [keras-team/keras](https://github.com/keras-team/keras/issues) repo.
To know more refer to:
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)

**Comment by EnderWiggin14** [comment]
@tilakrayal 

I just posted this to the Keras issues page.

**Comment by tilakrayal** [comment]
@EnderWiggin14 ,
Please feel free to move this issue to closed status as it has been tracked in Keras repo.Thanks!
